lib/charms/karma_k8s/v0/karma.py

```diff
@@ -0,0 +1,260 @@
+        return KarmaAlertmanagerConfig.from_dict({"name": name, "uri": uri})
+
+
+class KarmaAlertmanagerConfigChanged(EventBase):
+    def __init__(self, handle, data=None):
+        super().__init__(handle)
+        self.data = data
+
+    def snapshot(self):
+        """Save relation data."""
+        return {"data": self.data}
+
+    def restore(self, snapshot):
+        """Restore relation data."""
+        self.data = snapshot["data"]
+
+
+class KarmaProviderEvents(ObjectEvents):
+    alertmanager_config_changed = EventSource(KarmaAlertmanagerConfigChanged)
+
+
+class KarmaProvider(ProviderBase):
+    """A "provider" handler to be used by the Karma charm (the 'provides' side of the 'karma' relation).
+    This library offers the interface needed in order to provide Alertmanager URIs and associated information to the
+    Karma application.
+
+    To have your charm provide URIs to Karma, declare the interface's use in your charm's metadata.yaml file:
+
+    ```yaml
+    provides:
+      karma-dashboard:
+        interface: karma_dashboard
+    ```
+
+    A typical example of importing this library might be
+
+    ```python
+    from charms.alertmanager_karma.v0.karma import KarmaProvider
+    ```
+
+    In your charm's `__init__` method:
+
+    ```python
+    self.provider = KarmaProvider(
+        self, "karma-dashboard", "karma", "0.86"
+    )
+    ```
+
+    The provider charm is expected to observe and respond to the :class:`KarmaAlertmanagerConfigChanged` event,
+    for example:
+
+    ```python
+    self.framework.observe(
+        self.provider.on.alertmanager_config_changed, self._on_alertmanager_config_changed
+    )
+    ```
+
+    This provider observes relation joined, changed and departed events on behalf of the charm.
+
+    From charm code you can then obtain the list of proxied alertmanagers via:
+
+    ```python
+    alertmanagers = self.provider.get_alertmanager_servers()
+    ```
+
+    Arguments:
+            charm (CharmBase): consumer charm
+            name (str): relation name from consumer's metadata.yaml
+            service_name (str): service name (must be consistent the consumer)
+            version (str): semver-compatible version string
+
+    Attributes:
+            charm (CharmBase): consumer charm
+    """
+
+    on = KarmaProviderEvents()
+    _stored = StoredState()
+
+    def __init__(self, charm, name: str, service_name: str, version: str = None):
+        super().__init__(charm, name, service_name, version)
+        self.charm = charm
+        self._service_name = service_name
+
+        events = self.charm.on[self.name]
+        self.framework.observe(events.relation_joined, self._on_relation_joined)
+        self.framework.observe(events.relation_changed, self._on_relation_changed)
+        self.framework.observe(events.relation_departed, self._on_relation_departed)
+        self._stored.set_default(active_relations=set())
+
+    def get_alertmanager_servers(self) -> List[Dict[str, str]]:
```

Comment:
There is surprisingly no docstring for a public method. Knowing the return value is a "list of dictionaries with both keys and values as strings" still requires the client programmer to read the code or guess that the key **may** be a name and the value **may** be an address.

Reply:
docstings added here:
https://github.com/canonical/karma-operator/commit/fb49bd14172fd19e69a6a5170f4b6e42a47fc597